### PR TITLE
check type function for action callbacks

### PR DESCRIPTION
**Pull request by mum-never-proud**
I accidentally gave callback as a string rather than function, the app broke, I think its good to check action type is a function before calling it

### CODE REVIEW

**Comment by mum-never-proud** [comment]
@lukemelia do you think its better to warn as well?

**Comment by lukemelia** [comment]
@mum-never-proud Yes, warning seems appropriate

**Comment by mum-never-proud** [comment]
@lukemelia I have added warning I am not sure if the warning message is effective, let me know your thoughts

File: addon/components/modal-dialog.js [hunk 1]
```diff
@@ -110,18 +110,30 @@ export default Component.extend({
...
-        this.get('onClose')();
+      const onClose = this.get('onClose');
+      // we shouldn't warn if the callback is not provided at all
+      if (!isNone(onClose)) {
+        if (typeOf(onClose) === 'function') {
+          onClose();
+        } else {
+          warn('onClose handler must be a function', false, { id: 'ember-modal-dialog.on-close' })
+        }
       }
```
Comment (lukemelia) [inline addon/components/modal-dialog.js:121]:
```suggestion
      if (isNone(onClose)) {
        return;
      }
      
      if (typeOf(onClose) === 'function') {
        onClose();
      } else {
        warn('onClose handler must be a function', false, { id: 'ember-modal-dialog.on-close' })
      }
```

File: addon/components/modal-dialog.js [hunk 2]
```diff
@@ -110,18 +110,30 @@ export default Component.extend({
...
+      if (!isNone(onClickOverlay)) {
+        if (typeof(onClickOverlay) === 'function') {
+          onClickOverlay();
+        } else {
+          warn('onClickOverlay handler must be a function', false, { id: 'ember-modal-dialog.on-click-overlay' })
         }
+      } else {
+        // call close action
+        this.send('onClose');
       }
```
Comment (lukemelia) [inline addon/components/modal-dialog.js:137]:
```suggestion
      if (isNone(onClickOverlay)) {
        this.send('onClose');
        return;
      }
      if (typeof(onClickOverlay) === 'function') {
        onClickOverlay();
      } else {
        warn('onClickOverlay handler must be a function', false, { id: 'ember-modal-dialog.on-click-overlay' })
      }
```

**Comment by mum-never-proud** [comment]
@lukemelia yep I agree there's no point in proceeding if the callback function is not valid, in that case, can we use assert?

also, I am not sure how to test the exception handler in ember-qunit hints?

**Comment by lukemelia** [comment]
@mum-never-proud this has always been tricky -- I'd try https://github.com/workmanw/ember-qunit-assert-helpers

**Comment by mum-never-proud** [comment]
@lukemelia I am not sure why is this happening as not much expert in ember

unfortunately `assert.throws` or `assert.expectAssertion` doesn't support `async` (may be i am wrong)

even with `try-catch`, I am unable to swallow the error

```javascript
try {
  await click('#example-invalid-handler button');
  await click(overlaySelector);
} catch (err) {}
```

does this line

```javascript
assert('onClickOverlay handler must be a function', typeOf(onClickOverlay) === 'function');
```

has any special effect in ember? I hope its the regular `throw`

thanks for your patience 💪

**Comment by lukemelia** [comment]
@mum-never-proud I looked into it and tried a bunch of different approaches and was not successful in figuring out how to assert on an async Ember assert. All things considered, I am OK merging without this test coverage. Let me know what you think.

**Comment by mum-never-proud** [comment]
yes even with multiple attempts, testing async wasn't a success I will open an issue on `emberjs` for the same.

but for now, I am cool as well

I have done the manual testing and it works as anticipated, it should throw an error during development for invalid handlers

cc: @lukemelia

**Comment by lukemelia** [comment]
Thanks @mum-never-proud !

**Pull request merged**